File: packages/merkle-tree/src/test/sizeDeletion.test.ts
import * as crypto from 'crypto';
import { 
  IMerkleTree, 
  addItem, 
  deleteItem,
  findItemNode,
  createTree,
  HashedItem
} from '../lib/merkle-tree';

describe('Size calculation with file deletion', () => {
  
  // Helper function to create a file hash with specific size
  function createHashedItem(name: string, content: string, size: number): HashedItem {
    const hash = crypto.createHash('sha256')
      .update(content)
      .digest();
    return {
      name,
      hash,
      length: size,
      lastModified: new Date(),
    };
  }

  // Helper to build a test tree with files of specific sizes
  function buildTestTree(): IMerkleTree<any>{
    let tree = createTree("12345678-1234-5678-9abc-123456789abc");
    const files = [
      { name: 'file1.txt', content: 'content 1', size: 1000 },
      { name: 'file2.txt', content: 'content 2', size: 2000 },
      { name: 'file3.txt', content: 'content 3', size: 3000 },
      { name: 'file4.txt', content: 'content 4', size: 4000 },
      { name: 'file5.txt', content: 'content 5', size: 5000 }
    ];
    
    for (const file of files) {
      tree = addItem(tree, createHashedItem(file.name, file.content, file.size));
    }
    
    return tree;
  }

  test('deleting a file removes it completely from the tree', () => {
    const tree = buildTestTree();
    const fileToDelete = 'file3.txt';
    
    // Verify initial state
    const initialNode = findItemNode(tree, fileToDelete);
    expect(initialNode?.size).toBe(3000);
    
    // Delete the file completely
    deleteItem(tree, fileToDelete);
    
    // Verify the file is completely gone
    const deletedNode = findItemNode(tree, fileToDelete);
    expect(deletedNode).toBeUndefined();
  });

  test('deleting a file updates parent node sizes', () => {
    const tree = buildTestTree();
    const initialTotalSize = 15000; // Sum of all file sizes: 1000+2000+3000+4000+5000
    
    // Verify initial total size
    expect(tree.sort?.size).toBe(initialTotalSize);
    expect(tree.sort?.size).toBe(initialTotalSize);
    
    // Delete a file of size 3000
    deleteItem(tree, 'file3.txt');
    
    // Expected new total: 15000 - 3000 = 12000
    expect(tree.sort?.size).toBe(12000);
    expect(tree.sort?.size).toBe(12000);
  });

  test('deleting multiple files correctly updates size throughout the tree', () => {
    const tree = buildTestTree();
    const initialTotalSize = 15000;
    
    // Verify initial state
    expect(tree.sort?.size).toBe(initialTotalSize);
    
    // Delete multiple files
    deleteItem(tree, 'file1.txt'); // -1000
    deleteItem(tree, 'file4.txt'); // -4000
    
    // Expected new size: 15000 - (1000 + 4000) = 10000
    expect(tree.sort?.size).toBe(10000);
    expect(tree.sort?.size).toBe(10000);
    
    // Delete one more file
    deleteItem(tree, 'file5.txt'); // -5000
    
    // Expected final size: 10000 - 5000 = 5000
    expect(tree.sort?.size).toBe(5000);
    expect(tree.sort?.size).toBe(5000);
  });

  test('sizes are correctly propagated up through all parent nodes', () => {
    // Create a tree with 7 files to create a 3-level tree
    let tree = createTree("12345678-1234-5678-9abc-123456789abc");
    const fileSizes = [100, 200, 300, 400, 500, 600, 700];
    
    for (let i = 0; i < fileSizes.length; i++) {
      tree = addItem(tree, createHashedItem(`file${i}.txt`, `content ${i}`, fileSizes[i]));
    }
    
    // A 7-file tree will have a structure like:
    //          root
    //         /    \
    //     ABCD      EFG
    //    /    \    /   \
    //   AB    CD  EF    G
    //  / \   / \  / \
    // A   B C   D E  F
    
    // Verify initial state
    const totalSize = fileSizes.reduce((sum, size) => sum + size, 0);
    expect(tree.sort?.size).toBe(totalSize);
    
    // Delete file C (index 2, size 300)
    deleteItem(tree, 'file2.txt');
    
    // Verify file C is completely gone
    const nodeC = findItemNode(tree, 'file2.txt');
    expect(nodeC).toBeUndefined();
    
    // Verify root node size is updated
    expect(tree.sort?.size).toBe(totalSize - fileSizes[2]);
    expect(tree.sort?.size).toBe(totalSize - fileSizes[2]);
  });

  test('totalSize in metadata correctly reflects all deletions', () => {
    const tree = buildTestTree();
    const initialTotalSize = 15000;
    
    // Delete files one by one and verify metadata is updated each time
    deleteItem(tree, 'file1.txt'); // -1000
    expect(tree.sort?.size).toBe(initialTotalSize - 1000);
    
    deleteItem(tree, 'file3.txt'); // -3000
    expect(tree.sort?.size).toBe(initialTotalSize - 1000 - 3000);
    
    deleteItem(tree, 'file5.txt'); // -5000
    expect(tree.sort?.size).toBe(initialTotalSize - 1000 - 3000 - 5000);
    
    // Final size should be 2000 + 4000 = 6000
    expect(tree.sort?.size).toBe(6000);
  });
});

File: packages/merkle-tree-zig/src/test/sizeDeletion.test.zig
//
// Size calculation with file deletion (port of src/test/sizeDeletion.test.ts).
//

const std = @import("std");
const merkle_tree_zig = @import("merkle-tree-zig");
const merkle_verify = @import("merkle-verify.zig");
const merkle_tree = merkle_tree_zig.merkle_tree;
const IMerkleTree = merkle_tree.IMerkleTree;

//
// Helper to build a test tree with files of specific sizes
//
fn buildTestTree(allocator: std.mem.Allocator) !IMerkleTree {
    var tree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    const names = [_][]const u8{ "file1.txt", "file2.txt", "file3.txt", "file4.txt", "file5.txt" };
    const contents = [_][]const u8{ "content 1", "content 2", "content 3", "content 4", "content 5" };
    const sizes = [_]u64{ 1000, 2000, 3000, 4000, 5000 };
    for (names, contents, sizes) |name, content, size| {
        tree = try merkle_tree.addItem(allocator, &tree, try merkle_verify.createSha256HashedItem(allocator, name, content, size));
    }
    return tree;
}

test "deleting a file removes it completely from the tree" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = try buildTestTree(allocator);
    try std.testing.expectEqual(@as(u64, 3000), (merkle_tree.findItemInTree(tree.sort, "file3.txt")).?.size);
    try merkle_tree.deleteItem(allocator, &tree, "file3.txt");
    try std.testing.expect((merkle_tree.findItemInTree(tree.sort, "file3.txt")) == null);
}

test "deleting a file updates parent node sizes" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = try buildTestTree(allocator);
    try std.testing.expectEqual(@as(u64, 15000), tree.sort.?.size);
    try merkle_tree.deleteItem(allocator, &tree, "file3.txt");
    try std.testing.expectEqual(@as(u64, 12000), tree.sort.?.size);
}

test "deleting multiple files correctly updates size throughout the tree" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = try buildTestTree(allocator);
    try std.testing.expectEqual(@as(u64, 15000), tree.sort.?.size);
    try merkle_tree.deleteItem(allocator, &tree, "file1.txt"); // -1000
    try merkle_tree.deleteItem(allocator, &tree, "file4.txt"); // -4000
    try std.testing.expectEqual(@as(u64, 10000), tree.sort.?.size);
    try merkle_tree.deleteItem(allocator, &tree, "file5.txt"); // -5000
    try std.testing.expectEqual(@as(u64, 5000), tree.sort.?.size);
}

test "sizes are correctly propagated up through all parent nodes" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    const fileSizes = [_]u64{ 100, 200, 300, 400, 500, 600, 700 };
    for (fileSizes, 0..) |size, index| {
        const name = try std.fmt.allocPrint(allocator, "file{d}.txt", .{index});
        const content = try std.fmt.allocPrint(allocator, "content {d}", .{index});
        tree = try merkle_tree.addItem(allocator, &tree, try merkle_verify.createSha256HashedItem(allocator, name, content, size));
    }
    try std.testing.expectEqual(@as(u64, 2800), tree.sort.?.size);

    try merkle_tree.deleteItem(allocator, &tree, "file2.txt");
    try std.testing.expect((merkle_tree.findItemInTree(tree.sort, "file2.txt")) == null);
    try std.testing.expectEqual(@as(u64, 2800 - 300), tree.sort.?.size);
}

test "totalSize in metadata correctly reflects all deletions" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = try buildTestTree(allocator);
    try merkle_tree.deleteItem(allocator, &tree, "file1.txt");
    try std.testing.expectEqual(@as(u64, 14000), tree.sort.?.size);
    try merkle_tree.deleteItem(allocator, &tree, "file3.txt");
    try std.testing.expectEqual(@as(u64, 11000), tree.sort.?.size);
    try merkle_tree.deleteItem(allocator, &tree, "file5.txt");
    try std.testing.expectEqual(@as(u64, 6000), tree.sort.?.size);
}
